rclrs/src/node/graph.rs

```diff
@@ -0,0 +1,568 @@
+        topic: &str,
+    ) -> Result<Vec<TopicEndpointInfo>, RclrsError> {
+        self.get_publisher_subscriber_info_by_topic(topic, rcl_get_subscriptions_info_by_topic)
+    }
+
+    /// Get an rcl names_and_types function, with a "demangle" argument
+    fn get_names_and_types_by_node_with_demangle_arg(
+        &self,
+        node: &str,
+        namespace: &str,
+        getter: unsafe extern "C" fn(
+            *const rcl_node_t,
+            *mut rcl_allocator_t,
+            bool,
+            *const ::std::os::raw::c_char,
+            *const ::std::os::raw::c_char,
+            *mut rcl_names_and_types_t,
+        ) -> rcl_ret_t,
+    ) -> Result<TopicNamesAndTypes, RclrsError> {
+        let mut rcl_names_and_types = unsafe { rmw_get_zero_initialized_names_and_types() };
+        let node_name = CString::new(node).unwrap();
+        let node_namespace = CString::new(namespace).unwrap();
+        unsafe {
+            getter(
+                &*self.rcl_node_mtx.lock() as *const _,
+                &mut rcutils_get_default_allocator(),
+                false,
+                node_name.as_ptr(),
+                node_namespace.as_ptr(),
+                &mut rcl_names_and_types,
+            )
+        };
+
+        convert_names_and_types(rcl_names_and_types)
+    }
+
+    /// Get an rcl names_and_types function, without a "demangle" argument
+    fn get_names_and_types_by_node_without_demangle_arg(
+        &self,
+        node: &str,
+        namespace: &str,
+        getter: unsafe extern "C" fn(
+            *const rcl_node_t,
+            *mut rcl_allocator_t,
+            *const ::std::os::raw::c_char,
+            *const ::std::os::raw::c_char,
+            *mut rcl_names_and_types_t,
+        ) -> rcl_ret_t,
+    ) -> Result<TopicNamesAndTypes, RclrsError> {
+        let mut rcl_names_and_types = unsafe { rmw_get_zero_initialized_names_and_types() };
+        let node_name = CString::new(node).unwrap();
+        let node_namespace = CString::new(namespace).unwrap();
+        unsafe {
+            getter(
+                &*self.rcl_node_mtx.lock() as *const _,
+                &mut rcutils_get_default_allocator(),
+                node_name.as_ptr(),
+                node_namespace.as_ptr(),
+                &mut rcl_names_and_types,
+            )
+        };
+
+        convert_names_and_types(rcl_names_and_types)
+    }
+
+    /// Get publisher or subscriber info by topic
+    fn get_publisher_subscriber_info_by_topic(
+        &self,
+        topic: &str,
+        getter: unsafe extern "C" fn(
+            *const rcl_node_t,
+            *mut rcl_allocator_t,
+            *const ::std::os::raw::c_char,
+            bool,
+            *mut rcl_topic_endpoint_info_array_t,
+        ) -> rcl_ret_t,
+    ) -> Result<Vec<TopicEndpointInfo>, RclrsError> {
+        let topic = CString::new(topic).unwrap();
+        let mut rcl_publishers_info =
+            unsafe { rmw_get_zero_initialized_topic_endpoint_info_array() };
```

Comment:
Please add a `// SAFETY` comment to each `unsafe` block. E.g. for these `rmw_get_zero_initialized_*` calls, we've used

```
// SAFETY: Getting a zero-initialized value is always safe.
```
in other places.